pkg/generator/main.go

```diff
@@ -0,0 +1,158 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"fmt"
+	"log"
+	"os"
+	"regexp"
+	"strings"
+
+	apiext "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	"sigs.k8s.io/controller-tools/pkg/crd"
+	"sigs.k8s.io/controller-tools/pkg/loader"
+	"sigs.k8s.io/controller-tools/pkg/markers"
+	"sigs.k8s.io/yaml"
+)
+
+const (
+	ExperimentalMarker      = "<gateway:experimental>"
+	BundleVersionAnnotation = "gateway.networking.k8s.io/bundle-version"
+	ChannelAnnotation       = "gateway.networking.k8s.io/channel"
+)
+
+// This generation code is largely copied from sigs.k8s.io/controller-tools/pkg/crd/gen.go
+func main() {
+	roots, err := loader.LoadRoots(
+		"k8s.io/apimachinery/pkg/runtime/schema", // Needed to parse generated register functions.
+		"sigs.k8s.io/gateway-api/apis/v1alpha2",
+	)
+
+	if err != nil {
+		log.Fatalf("failed to load package roots: %s", err)
+	}
+
+	generator := &crd.Generator{}
+
+	parser := &crd.Parser{
+		Collector: &markers.Collector{Registry: &markers.Registry{}},
+		Checker: &loader.TypeChecker{
+			NodeFilters: []loader.NodeFilter{generator.CheckFilter()},
+		},
+	}
+
+	err = generator.RegisterMarkers(parser.Collector.Registry)
+	if err != nil {
+		log.Fatalf("failed to register markers: %s", err)
+	}
+
+	crd.AddKnownTypes(parser)
+	for _, r := range roots {
+		parser.NeedPackage(r)
+	}
+
+	metav1Pkg := crd.FindMetav1(roots)
+	if metav1Pkg == nil {
+		log.Fatalf("no objects in the roots, since nothing imported metav1")
+	}
+
+	kubeKinds := crd.FindKubeKinds(parser, metav1Pkg)
+	if len(kubeKinds) == 0 {
+		log.Fatalf("no objects in the roots")
+	}
+
+	channels := []string{"stable", "experimental"}
+	for _, channel := range channels {
+		for groupKind := range kubeKinds {
+			log.Printf("generating %s CRD for %v\n", channel, groupKind)
+
+			parser.NeedCRDFor(groupKind, nil)
+			crdRaw := parser.CustomResourceDefinitions[groupKind]
+
+			// Inline version of "addAttribution(&crdRaw)" ...
+			if crdRaw.ObjectMeta.Annotations == nil {
+				crdRaw.ObjectMeta.Annotations = map[string]string{}
+			}
+			crdRaw.ObjectMeta.Annotations[BundleVersionAnnotation] = "v0.5.0-dev"
```

Comment:
Not a biggie but if you can pull this out of code into the script invoking this code, that may help as we do releases.

Reply:
I guess we will need to also change the PR URL for every release as well. Your call.

Reply:
I wasn't sure where to place these. I thought about another package, but wasn't sure we wanted to commit to exporting them. For now I've just moved these up to the top in a `const` block. My current thought process is that these would be changed as part of every changelog PR, and then immediately changed back in a follow up PR post release. There's likely a better way to handle this, but I think that approach would technically work. Trying not to make this PR much bigger, but will work on a follow up for docs.